scripts/ci-build.sh

```diff
@@ -0,0 +1,512 @@
+#!/bin/bash -e
+
+# (C) 2018 GENIVI Alliance
+# LICENSE: MPLv2
+
+# ---- General settings ----
+
+# Fixed cache location on all agents that support this.
+AGENT_STANDARD_DL_DIR="/var/cache/yocto/downloads"
+AGENT_STANDARD_SSTATE_DIR="/var/cache/yocto/sstate"
+AGENT_STANDARD_SGX_LOCATION="/var/go/sgx_bin"
+AGENT_STANDARD_SGX_GEN3_LOCATION="/var/go/sgx_bin_gen3"
+
+# ---- Helper functions ----
+
+fail=false
+
+# This is some trickery, but quite useful.
+# Return the value of a variable whose name is stored in another variable
+deref() { eval echo \$$1 ; }
+
+ensure_var_is_defined() {
+  var_name=$1
+  current_value=$(deref $var_name)
+  if [[ -z "$current_value" ]]; then
+    echo "Please define environment var_name \$$var_name (found empty value)"
+    fail=true  # We want all the errors reported, so don't halt yet.
+  fi
+}
+
+define_with_default() {
+  var_name=$1
+  default="$2"
+  current_value=$(deref $var_name)
+  if [[ -z "$current_value" ]]; then
+    eval $var_name="$default"
+  fi
+}
+
+stop_if_failure() {
+  if [[ "$fail" == "true" ]] ; then
+    exit 1
+  fi
+}
+
+stop_immediately() {
+  echo "Fatal error occurred - stopping ci-build script"
+  exit 2
+}
+
+append_local_conf() {
+  LOCAL_CONF="$BASEDIR/build/conf/local.conf"
+  if [[ -f "$LOCAL_CONF" ]]; then
+    echo -n "Appending to local.conf: "
+    cat <<EOT | tee -a "$LOCAL_CONF"
+$1
+EOT
+  else
+    echo "Fatal: Did not find local.conf where expected"
+    stop_immediately
+  fi
+}
+
+stage_artifact() {
+  mkdir -p staging
+  cmd=$1
+  shift
+
+  if [[ -z "$1" ]] ; then
+    echo "Skipped an artifact pattern that matched nothing"
+    return
+  fi
+
+  for f in $@ ; do  # <- could be empty glob list
+    flag=
+    if [[ -e "$f" ]] ; then
+      if [[ $cmd == cp && -d "$f" ]] ; then
+        flag=-a
+      elif [[ $cmd == cp ]] ; then
+        flag=--preserve=all
+      fi
+      $cmd $flag "$f" staging/
+    else
+      echo "Skipping non-existing (or already moved) artifact: $f"
+    fi
+  done
+}
+
+# Layer version overrides:
+# Find and evaluate environment variables of type
+
+# LAYER_<layername>_FORK
+# LAYER_<layername>_COMMIT
+# LAYER_<layername>_BRANCH
+# LAYER_<layername>_TAG
+
+# *NOTE*  Variables cannot have dashes in name therefore, underscores must be
+# used!  For example: LAYER_meta_openembedded for meta-openembedded layer
+
+# If an override variable is defined, build with this version of the layer
+# instead.  If multiple conflicting values, then the most specific override
+# (commit) wins
+
+layer_override() {
+  local name=$1 suffix var_name value d
+
+  # Variables cannot have dashes in name therefore, underscores must be
+  # used.  For example: LAYER_meta_browser_COMMIT for meta-browser layer
+  clean_name=$(echo "$name" | sed 's/-/_/g')
```

Comment:
Could use string substitution directly in bash, it's just that I'm familiar with the sed syntax...